openfe/protocols/openmm_rfe/equil_rfe_methods.py

```diff
@@ -637,57 +637,71 @@ def run(self, *, dry=False, verbose=True,
 
-            sampler.minimize(max_iterations=sim_settings.minimization_steps)
+                sampler.minimize(max_iterations=sim_settings.minimization_steps)
 
-            # equilibrate
-            if verbose:
-                logger.info("equilibrating systems")
+                # equilibrate
+                if verbose:
+                    logger.info("equilibrating systems")
 
-            sampler.equilibrate(int(equil_steps.m / mc_steps))  # type: ignore
+                sampler.equilibrate(int(equil_steps.m / mc_steps))  # type: ignore
 
-            # production
-            if verbose:
-                logger.info("running production phase")
+                # production
+                if verbose:
+                    logger.info("running production phase")
 
-            sampler.extend(int(prod_steps.m / mc_steps))  # type: ignore
+                sampler.extend(int(prod_steps.m / mc_steps))  # type: ignore
 
-            # calculate estimate of results from this individual unit
-            ana = multistate.MultiStateSamplerAnalyzer(reporter)
-            est, _ = ana.get_free_energy()
-            est = (est[0, -1] * ana.kT).in_units_of(omm_unit.kilocalories_per_mole)
-            est = ensure_quantity(est, 'openff')
+                # calculate estimate of results from this individual unit
+                ana = multistate.MultiStateSamplerAnalyzer(reporter)
+                est, _ = ana.get_free_energy()
+                est = (est[0, -1] * ana.kT).in_units_of(omm_unit.kilocalories_per_mole)
+                est = ensure_quantity(est, 'openff')
 
-            # close reporter when you're done
+                nc = shared_basepath / sim_settings.output_filename
+                chk = shared_basepath / sim_settings.checkpoint_storage
+            else:
+                # clean up the reporter file
+                fns = [shared_basepath / sim_settings.output_filename,
+                       shared_basepath / sim_settings.checkpoint_storage]
+                for fn in fns:
+                    os.remove(fn)
+        finally:
+            # close reporter when you're done, prevent file handle clashes
             reporter.close()
 
-            nc = shared_basepath / sim_settings.output_filename
-            chk = shared_basepath / sim_settings.checkpoint_storage
+            # clear GPU contexts
+            #energy_context_cache.empty()
+            #sampler_context_cache.empty()
+            # TODO: remove once upstream solution in place: https://github.com/choderalab/openmmtools/pull/690
+            # replace with above
+            for context in list(energy_context_cache._lru._data.keys()):
+                del energy_context_cache._lru._data[context]
+            for context in list(sampler_context_cache._lru._data.keys()):
+                del sampler_context_cache._lru._data[context]
+
+            del sampler_context_cache, energy_context_cache
+            del integrator, sampler
```

Comment:
We can't do this here - if we delete the sampler we can't return it on line 691

Reply:
Ah good catch!